### ISSUE TITLE
MultiAppNearestNodeTransfer gives wrong results in parallel

### ISSUE THREAD

**Issue by bwspenc** [idaholab/moose#8004]
### Description of the enhancement or error report
Yingjie Liu posted an issue to the mailing list a couple of times about a problem with the MultiAppNearestNodeTransfer in parallel.  He's running into trouble transferring from a surface to a volume.  The master model is a 2D mesh, and the submodel is a 1D mesh. It works fine in serial, but in parallel with a sufficiently large mesh, the results transferred to the master 2D mesh get corrupted.

### Rationale for the enhancement or information for reproducing the error
I am checking in his example problem into the directory 
moose/test/tests/transfers/multiapp_nearest_node_transfer_bug on the nn_transfer_bug branch on my git repository.  It's a pretty small problem based on a moose test problem, and you can run it with the moose_test app.  It starts giving bad results for me with a mesh (the 2D master app mesh) of around 100x100 or larger on 2 processors.

### Identified impact
This is probably a bug within MultiAppNearestNodeTransfer

**Comment by bwspenc** [comment]
@yingjie1989

**Comment by permcody** [comment]
@yingjie1989, Thanks for the report. What are you working on? I'm trying to figure out what kind of priority to assign to this ticket. Yes this is bad but we're all pretty busy. If this is a road block for you, I can raise the priority.

**Comment by bwspenc** [comment]
This is a roadblock for him. I was planning to take a look at it, but I'm pretty busy too. I gave him a few tips on how to get started debugging it himself.

**Comment by bwspenc** [comment]
@permcody I tried undoing that most recent change to the transfer, but still see the same issue, so it's something else. I'm still looking into this a little more, and will let you know if I find anything.

**Comment by yingjie1989** [comment]
@permcody This is really a roadblock for me. I am working on a coupling problem which needs the data transfer between a 3D domain and 2D domain.

**Comment by friedmud** [comment]
@yingjie1989 I'm very sorry.  This has been on my TODO list for a couple of weeks now... but I've been swamped.

I'm finally free a bit this weekend though so I should be able to get to it.

@bwspenc Thanks for getting the ball rolling here.  Let us know if you find anything and I'll help out this weekend.

**Comment by yingjie1989** [comment]
@permcody I tested a new problem in which data are transferred between a 2D domain and a 1D domain. The data of the 1D domain should be transferred to the top boundary of the 2D domain. No difference should be observed whether I set the "target_boundary = true" or not. But that is not the case.

**Comment by bwspenc** [comment]
@friedmud I had a little time to look at it this afternoon, but didn't come up with anything, and I'm heading out now.  I'll pick this back up again on Monday if I don't hear from you.

**Comment by friedmud** [comment]
Sounds good - thanks for taking a look!

**Comment by friedmud** [comment]
Ok - I have this built - and running @bwspenc 's problem I can see that the output is wrong.

I have no idea what could be causing this.  I'm in class now... but when I get out this is my # 1 priority this afternoon.

**Comment by friedmud** [comment]
I think I know what the issue is.  Unfortunately... it's a HUGE problem that may have been affecting MANY answers over the last year and a half....

Look for an answer shortly...

**Comment by friedmud** [comment]
Well.  Crap.

You can see the fix here: https://github.com/friedmud/moose/commit/56c6932f22ca6f62de5805aa9178425875c24c67

**Good news**: the fix is easy

**Bad news**: it's been potentially causing corrupt garbage to get transferred in _every calculation_ that's been done using `MultiAppNearestNodeTransfer` (and I suspect other Transfers) for a year and half.

The issue is this: a non-blocking send was being used... but the buffers being sent were going out of scope and getting destroyed before the send was known to be complete.  The buffer being sent MUST live until the send is complete (which you can check using `MPI_test()` or by waiting using `MPI_wait()` which happens at the end of the Transfer).

This was introduced by @smharper here: https://github.com/idaholab/moose/pull/5576

It looks like it definitely also affects `MultiAppMeshFunctionTransfer`... and I haven't checked other Transfers yet.

I have no idea how this wasn't caught until now.  Any largish problem should have shown the issue.

It won't get caught by small problems because... MPI.  MPI does this thing where if a buffer is really small then it will just copy it locally into a "cache" space... and then do the send from there.  So for small problems MPI was "fixing" the problem for us.  But that space is not very large... we definitely would have been hitting this on any larger problems.... this could manifest in _many_ ways including non-convergence of Picard iterations, etc.  Just nasty.

**Comment by friedmud** [comment]
BTW: A big thanks to @smharper for helping me track this down tonight.  After 1.5 years he was able to remember working on the code and he got me pointed in the right direction to spot the problem.

**Comment by friedmud** [comment]
I fixed up `MultiAppMeshFunctionTransfer`... and grepped though other Transfers and didn't find the same problem.

Interestingly... the same pattern was in `MultiAppProjectionTransfer`... but it had been done _correctly_ there.  Not quite sure why it right in one place and not in the other two!

**Comment by friedmud** [comment]
Can we get everyone to take a look through these Transfers and just make sure we don't have any more of this?

If we can't find any more of them I'll put in a PR with the new test for `MultiAppNearestNodeTransfer`.  I wonder if I can make a test that will cause `MultiAppMeshFunctionTransfer` to fail too...

**Comment by bwspenc** [comment]
@friedmud and @smharper, thanks a lot for tracking this down!  Maybe we could have some tests for these transfers that have larger meshes where we don't do a solve so they can still run fast. Just populate fields with AuxKernels or something and transfer them. I guess there would be no guarantee we'd catch everything, but we'd have better coverage than we do now with just small regression tests.

**Comment by yingjie1989** [comment]
@bwspenc @permcody @friedmud @smharper Thank you all for helping me fix this issue, which makes it possible for me to push my research forward. I have pulled down the code and applied it to my App. I will report here if I have some new findings. Thank you all!

**Comment by permcody** [comment]
This is well into the land of undefined behavior and we are lucky that it was this easy to figure out that problem. I won't have any better feeling that things are working with larger tests. We need something better.  I'm hoping to start using this:
http://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html

**Comment by bwspenc** [comment]
By the way, @friedmud, while you're in there making changes, I noticed that `MultiAppNearestNodeTransfer::getNearestNode()` isn't used at all, and should probably be deleted.

**Comment by friedmud** [comment]
Actually - Valgrind should have caught this - if we were to run parallel
valgrind.  The bug here is accessing already freed memory... which is what
Valgrind is good at detecting.

I think the test that @bwspenc created is good enough to keep this bug from
resurfacing.  I'm going to try to make another one that shows the issue
with MultiAppMeshFunctionTransfer too.  It shouldn't be hard.

But: I agree with @permcody that what we really need is to be running tools
that can detect this... like parallel Valgrind... or Clang Address
Sanitizer.

**Comment by permcody** [comment]
@friedmud - you never made a PR out of this commit. Can you do that?

**Comment by friedmud** [comment]
Yep: I was waiting for @yingjie1989 to report if it worked or not... but I think it must have ;-)

**Comment by yingjie1989** [comment]
@friedmud Sorry for not reporting this earlier. I did not know you had been waiting. The modified nearestnodetransfer code works. Thanks

**Comment by friedmud** [comment]
Thanks @yingjie1989 !  I appreciate it!

Don't worry about me... I've got a million things going on anyway - you didn't hold me up.

**Comment by lindsayad** [comment]
It looks to me like there was never a Pull Request submitted for this?

**Comment by lindsayad** [comment]
Yep, Derek's fix resolved a corrupted transfer for me

**Comment by yingjie1989** [comment]
Derek’s fix helped me a lot. I think this fix definitely should be merged into the latest version of MOOSE if it does not induce other issues.
%%%%%%%%%%%%%%%%%%%
Yingjie Liu
PhD Candidate
Dept. Civil & Environ. Eng.
Information & Initiative @ Duke
Co-Chair of Students Advocating for Graduate Education
Duke University, Durham, NC27705
Email: yingjie.liu@duke.edu<mailto:yingjie.liu@duke.edu>

On Jun 23, 2017, at 1:50 PM, Alex Lindsay <notifications@github.com<mailto:notifications@github.com>> wrote:


Yep, Derek's fix resolved a corrupted transfer for me

**Comment by permcody** [comment]
@lindsayad - perhaps you can push up a PR? You should be able to keep Derek's ownership on the fix plus whatever else you might add to this PR (a new test would be nice).

**Comment by lindsayad** [comment]
I opened #9380. Will look at the test failure shortly
